parsers/CA_NS.py

```diff
@@ -1,213 +1,182 @@
-            if absolute_bound and value > absolute_bound:
-                valid = False
-                logger.warning(
-                    "discarding datapoint at {dt} due to {fuel} "
-                    "too high: {value} MW".format(
-                        dt=data_date, fuel=gen_type, value=value
-                    ),
-                    extra={"key": zone_key},
-                )
-        if not valid:
-            # continue the outer loop, not the inner
-            continue
-
-        production.append(
-            {
-                "zoneKey": zone_key,
-                "datetime": data_date,
-                "production": {
-                    key: value
-                    for key, value in electricity_mix.items()
-                    if key != "imports"
-                },
-                "source": "nspower.ca",
-            }
+        # In this source, imports are positive. In the expected result for
+        # CA-NB->CA-NS, "net" represents a flow from NB to NS, i.e., an import
+        # to NS, so the value can be used directly. Note that this API only
+        # specifies imports; when NS is exporting energy, the API returns 0.
+        exchanges.append(
+            datetime=timestamp,
+            netFlow=load * mix["Imports"] / 100,
+            source=SOURCE,
+            zoneKey=ZoneKey("CA-NB->CA-NS"),
         )
 
-        # In this source, imports are positive. In the expected result for CA-NB->CA-NS,
-        # "net" represents a flow from NB to NS, that is, an import to NS.
-        # So the value can be used directly.
-        # Note that this API only specifies imports. When NS is exporting energy, the API returns 0.
-        imports.append(
-            {
-                "datetime": data_date,
-                "netFlow": electricity_mix["imports"],
-                "sortedZoneKeys": "CA-NB->CA-NS",
-                "source": "nspower.ca",
-            }
+        production_mix = ProductionMix()
+        production_mix.add_value("biomass", load * mix["Biomass"] / 100)
+        production_mix.add_value("coal", load * mix["Solid Fuel"] / 100)
+        production_mix.add_value("gas", load * mix["CT's"] / 100)
+        production_mix.add_value("gas", load * mix["HFO/Natural Gas"] / 100)
+        production_mix.add_value("gas", load * mix["LM 6000's"] / 100)
+        production_mix.add_value("hydro", load * mix["Hydro"] / 100)
+        production_mix.add_value("wind", load * mix["Wind"] / 100)
+        # Sanity checks: verify that reported production doesn't exceed listed
+        # capacity by a lot. In particular, we've seen error cases where hydro
+        # production ends up calculated as 900 MW which greatly exceeds known
+        # capacity of 418 MW.
+        if (
+            100 < production_mix.biomass
+            or 1300 < production_mix.coal
+            or 700 < production_mix.gas
+            or 500 < production_mix.hydro
+            or 700 < production_mix.wind
```

Comment:
Since these can potentially be none this is throwing a bunch of type errors. We shoud use getters here instead with a fallback value or check that all are valid first.

Something like all() not None could work?

Reply:
Hmm, that's a good catch. Do you know which linter emits that warning? I don't see it when I run `poetry run check`.